### PR TITLE
[tunix] Add model-agnostic JAX diffusion contracts

### PR DESCRIPTION
**Pull request by ethannnnnn**
## Motivation

Diffusion training integrations need a small, stable JAX boundary that does
not depend on MaxText, a specific model, or a training algorithm.

## Scope

This PR adds only:

- a batch-major `DiffusionTokenBatch` with target-aligned token IDs and
  explicit loss weights;
- typed raw-batch adapter and logits-function protocols;
- shape, dtype, eager-value, and scorer-output validation; and
- focused contract tests and package documentation.

It does not add an SFT or distillation objective, change gradient reduction,
prepare a rollout, alter RL behavior, or import MaxText.

## Design

Every model-input leaf has the same leading batch axis. Static model
configuration remains in the logits-function closure rather than becoming a
batch leaf. Targets refer to physical prediction positions, so model-specific
same-position or shifted-logit alignment happens before constructing the
canonical batch.

`DiffusionLogitsFn` receives an NNX model and prepared model inputs and returns
floating-point logits with shape `[batch, length, vocabulary]`. Construction
and scoring boundaries validate array structure, target alignment, and active
values. Static checks remain safe under JIT and sharding; concrete finite and
range checks run when arrays are eagerly addressable.

## Compatibility

The package is new and opt-in. Existing SFT, distillation, RL, and
autoregressive paths do not construct these contracts and remain unchanged.

## Performance and Quality

This contracts-only PR does not add a diffusion training or inference runtime,
so it has no direct performance or quality delta relative to the existing
autoregressive path. The package README now compares their execution and
quality tradeoffs and defines the evaluation protocol for implementation PRs:
matched model, data, training-compute, hardware, batch, sequence, and decoding
budgets; latency and throughput for performance; and fixed tasks and seeds for
quality. It also calls out that autoregressive perplexity and diffusion
training loss are not directly interchangeable.

## Tests

```text
python -m pytest -q tests/diffusion/types_test.py
15 passed
```

Pyink, isort, scoped Pylint, Python compilation, and `git diff --check` also
pass for the changed files.

## Follow-Ups

Denominator-aware weighted reduction will be proposed separately after this
contract lands. Diffusion SFT, prepared distillation, OPD, rollout transport,
policy scoring, and GRPO integration remain later independent reviews.

Design document:
https://docs.google.com/document/d/1Xe-98ScS2RSH29AdhTdc9tO4WHQG5wCbFkJA3gmIUZQ/edit

### CODE REVIEW

File: tunix/diffusion/types.py [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+import jax
+import jax.numpy as jnp
+import numpy as np
+
+Array: TypeAlias = jax.Array | np.ndarray
+ModelInputs: TypeAlias = Mapping[str, Any]
+
+
+def _shape_and_dtype(name: str, value: Any) -> tuple[tuple[int, ...], Any]:
+  if not isinstance(value, (jax.Array, jax.core.Tracer, np.ndarray)):
```
Comment (wang2yn84) [inline tunix/diffusion/types.py]: Should consider jax.ShapeDtypeStruct if run jax.eval_shape()

File: tunix/diffusion/__init__.py [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+from tunix.diffusion.interfaces import compute_diffusion_logits
+from tunix.diffusion.interfaces import DiffusionBatchAdapter
+from tunix.diffusion.interfaces import DiffusionLogitsFn
+from tunix.diffusion.types import DiffusionTokenBatch
+
+__all__ = [
+    "DiffusionBatchAdapter",
+    "DiffusionLogitsFn",
+    "DiffusionTokenBatch",
+    "compute_diffusion_logits",
```
Comment (wang2yn84) [inline tunix/diffusion/__init__.py:29]: The design do explicitly lists validate_diffusion_logits as one of the primary public API functions. Consider add here

File: tunix/diffusion/types.py [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+from collections.abc import Mapping
+from typing import Any, TypeAlias
+
+import flax
+import jax
+import jax.numpy as jnp
+import numpy as np
+
+Array: TypeAlias = jax.Array | np.ndarray
+ModelInputs: TypeAlias = Mapping[str, Any]
```
Comment (wang2yn84) [inline tunix/diffusion/types.py]: Consider  jaxtyping.PyTree based on the doc

File: tunix/diffusion/interfaces.py [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+  """Converts an external batch to Tunix's canonical diffusion batch."""
+
+  def __call__(self, batch: RawBatchT_contra, /) -> types.DiffusionTokenBatch:
+    ...
+
+
+class DiffusionLogitsFn(Protocol):
+  """Computes target-aligned token logits for a diffusion model."""
+
+  def __call__(
```
Comment (wang2yn84) [inline tunix/diffusion/interfaces.py:36]: It'll be useful if you can put a docstring for __call__ function too. You already have compute_diffusion_logits, not sure what __call__ does.

File: tunix/diffusion/types.py (hunk 1 above)
Comment (ethannnnnn) [reply tunix/diffusion/types.py]: Done. `Array` and `_shape_and_dtype` now accept `jax.ShapeDtypeStruct`, and abstract values skip eager host checks. A regression test round-trips a `DiffusionTokenBatch` through `jax.eval_shape` and validates abstract logits.

File: tunix/diffusion/__init__.py (hunk 2 above)
Comment (ethannnnnn) [reply tunix/diffusion/__init__.py:29]: Done. `validate_diffusion_logits` is now imported and listed in `tunix.diffusion.__all__`; the abstract-logits test calls it through the public package surface.

File: tunix/diffusion/types.py (hunk 3 above)
Comment (ethannnnnn) [reply tunix/diffusion/types.py]: Done. `ModelInputs` is now `jaxtyping.PyTree[Array]`, the mapping-only runtime restriction is removed, and a nested tuple/dict PyTree test covers the generic contract.

File: tunix/diffusion/interfaces.py (hunk 4 above)
Comment (ethannnnnn) [reply tunix/diffusion/interfaces.py:36]: Done. Both protocol `__call__` methods now have method-level docstrings. `DiffusionLogitsFn.__call__` is documented as the model-specific scorer returning unvalidated target-aligned `[batch, length, vocabulary]` logits; `compute_diffusion_logits` owns validation.

**Comment by ethannnnnn** [comment]
Thanks for the suggestion. I added a performance and quality comparison with the existing autoregressive path to both the PR description and `tunix/diffusion/README.md`. Because this PR defines contracts only and does not add a diffusion training or inference runtime, it has no direct benchmark delta. The new section instead documents the execution and quality tradeoffs and the matched evaluation protocol that implementation PRs should use, without claiming unsupported gains.

**Review by wang2yn84 — APPROVED** [review]
